Интерфейс

Bubliks opened the pull request:
- добавлено меню
- добавлены разделы (меню, выбор игры, об игре, ввод имени пользователя, статистика игры)
- доблавлена таблица статистики
- добавлена графика
- добавлен основной цикл игры
- доблавлен обработчик событий нажатия кнопки мыши

Unatart committed:
~

Bubliks committed:
Добавлен интерфейс.

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+#include <string>
+
+enum COLOR_ID {
+    WHITE = '#',
```
это должно быть не в mainwindow.cpp - это основное замечание - надо разделить на модули...
в частности, этот enum скорее всего должен быть на уровне модели...

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+    size_map_width = 1;
+    size_map_height = 1;
+
+    size_canvas_width = 400;
```
магические константы
используйте конфигурационные файлы

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+QPoint MainWindow::getPosition(QPointF cursor) {
+    for (int i = 0; i < size_map_height; ++i) {
+        for (int j = 0; j < size_map_width; ++j) {
+            if (cursor.x() >= j * size_cube_width
```
а зачем цикл?
разве нельзя определить это остатком от деления?

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+    }
+}
+
+void MainWindow::drawCube(QPoint cube, int clr_pen, QColor clr_brush, float percent_offset) {
```
возможно, имеет смысл использовать один метод с некоторыми параметрами по умолчанию..

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+    delete ui;
+}
+
+QColor MainWindow::handleColor(int num_clr) {
```
лучше использовать enum class вместо int...

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+void MainWindow::loadData() {
+    //there are methods don`t work (rewrite in FileManager)
+    field = new PlayingField();
+    std::string filename = "maps/1.txt";
```
тут что-то не так
это либо должно читаться из GUI, либо вызываться не здесь и использовать конфигурационные файлы..

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+
+void MainWindow::start(){
+    //#########for fun#############
+    QSound sound("sound/background.wav");
```
используйте конфигурацию...

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+
+void MainWindow::on_Exit_clicked()
+{
+    exit(1);
```
а это не приведёт к разного рода утечкам в других модулях?

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+    toPage(ABOUT);
+}
+
+void MainWindow::on_Statistic_clicked()
```
а почему сверху все методы были в CamelCase, а тут - в смешанном стиле?

leshiy1295 commented on mainwindow.h:
```diff
@@ -0,0 +1,112 @@
+#ifndef MAINWINDOW_H
+#define MAINWINDOW_H
+
+#include <QSound>
```
возможно, тоже стоит выделить обобщённый интерфейс, который не зависит от Qt, и реализовать одного наследника через Qt...

leshiy1295 commented on mainwindow.h:
```diff
@@ -0,0 +1,112 @@
+    void on_backSelect_clicked();
+
+private:
+    Ui::MainWindow *ui;
```
можно заиспользовать умные указатели, чтобы не париться по поводу delete...

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,333 @@
+    drawCube(move_cube, clr, GREEN, 5);
+}
+
+QPoint MainWindow::ConvertToQPoint(Coordinates coord) {
```
вот этот слой по идее не должен быть в модуле рисования - возможно, это должен быть какой-то промежуточный слой-преобразователь.. но тут ещё не так принципиально..

leshiy1295 reviewed (changes requested):
Когда поправите - оставьте комментарий

Bubliks committed:
add sound, draw

Добавлен кастомный класс звука, отрисовки
Разбито на отдельные модули.
Добавлены умные указатели
Выделены конвертации точек в отдельный файл

Bubliks commented:
Готово к показу

leshiy1295 commented on customsound.cpp:
```diff
@@ -0,0 +1,40 @@
+void CustomSound::playInMenu() {
+    //playList->addMedia(QUrl(QString::fromStdString(soundPath) + "menu/background"));
+    stop();
+    setMedia(QUrl::fromLocalFile(QString::fromStdString(soundPath) + "menu/background.wav"));
```
это тоже должно быть в конфигурации

leshiy1295 commented on main.cpp:
```diff
@@ -0,0 +1,17 @@
+    QApplication a(argc, argv);
+
+    ConfiguratonWindow conf;
+    std::string filename =  "config.txt";
```
возможно, стоит брать это имя по умолчанию, но также проверить и argv

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,228 @@
+    ui->label_ver->setText(QString::fromStdString(conf.version_program));
+
+    //gameLoop
+    gameLoop = std::move(std::unique_ptr<QEventLoop>(new QEventLoop(ui->graphicsView)));
```
используйте make_unique

leshiy1295 commented on mainwindow.h:
```diff
@@ -0,0 +1,96 @@
+    void on_backSelect_clicked();
+
+private:
+    Ui::MainWindow *ui;
```
почему всё же Вы не используете умные указатели тут?

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,228 @@
+    drawPossibleCube();
+}
+
+void MainWindow::loadData(std::string map) {
```
это должна быть отдельная сущность...
У Вас сейчас MainWindow делает много чего - надо разнести по разным объектам

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,228 @@
+
+void MainWindow::runGame() {
+    //GETS AND SAVE MAP IN FILEMANAGER
+    std::vector<std::string> map = {"1.txt"}; //this example
```
1.txt должен браться откуда-то?

leshiy1295 commented on mainwindow.cpp:
```diff
@@ -0,0 +1,228 @@
+    //GETS AND SAVE MAP IN FILEMANAGER
+    std::vector<std::string> map = {"1.txt"}; //this example
+    for (size_t i = 0; i < map.size(); ++i) {
+        loadData(map[i]);
```
у Вас сейчас интерфейсный класс отвечает за подгрузку моделей - это неправильно

leshiy1295 reviewed (changes requested):
Не до конца разделили интерфейс и модель на уровне инициализации карты, а также не все пути/константы вынесли в конфигурацию/в интерфейс.
Когда поправите - отпишитесь
